File: java/1094_CarPooling.java
class Solution {
    public boolean carPooling(int[][] trips, int capacity) {
        int[] buckets = new int[1001];
        for (int[] trip : trips){
            buckets[trip[1]] += trip[0]; // beginning of carrying passengers
            buckets[trip[2]] -= trip[0]; // end of carrying passengers
        }
        
        int totalCurrentCapacity = 0;
        for (int numPeople : buckets) {
            totalCurrentCapacity += numPeople;
            if (totalCurrentCapacity > capacity)
                return false;
        }
        return true;
    }
}

File: cpp/1094_CarPooling.cpp
class Solution {
public:
    bool carPooling(vector<vector<int>>& trips, int capacity) {
        vector<int> buckets(1001, 0);
        for (vector<int> trip : trips) {
            buckets[trip[1]] += trip[0];
            buckets[trip[2]] -= trip[0];
        }
        int curr_capacity = 0;
        for (int n : buckets) {
            curr_capacity += n;
            if (curr_capacity > capacity)
                return false;
        }
        return true;
    }
};
